chromium_src/third_party/blink/renderer/core/frame/location.cc

```diff
@@ -0,0 +1,52 @@
+/* Copyright (c) 2023 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+#include "third_party/blink/renderer/core/frame/location.h"
+
+#include "third_party/blink/renderer/bindings/core/v8/binding_security.h"
+#include "third_party/blink/renderer/bindings/core/v8/v8_binding_for_core.h"
+#include "third_party/blink/renderer/core/dom/document.h"
+#include "third_party/blink/renderer/core/frame/csp/content_security_policy.h"
+#include "third_party/blink/renderer/core/frame/dom_window.h"
+#include "third_party/blink/renderer/core/frame/local_dom_window.h"
+#include "third_party/blink/renderer/core/frame/local_frame.h"
+#include "third_party/blink/renderer/core/frame/remote_dom_window.h"
+#include "third_party/blink/renderer/core/loader/frame_load_request.h"
+#include "third_party/blink/renderer/core/loader/frame_loader.h"
+#include "third_party/blink/renderer/core/url/dom_url_utils_read_only.h"
+#include "third_party/blink/renderer/platform/bindings/exception_state.h"
+#include "third_party/blink/renderer/platform/bindings/v8_dom_activity_logger.h"
+#include "third_party/blink/renderer/platform/bindings/v8_dom_wrapper.h"
+#include "third_party/blink/renderer/platform/weborigin/kurl.h"
+#include "third_party/blink/renderer/platform/weborigin/security_origin.h"
+
+#define Location Location_ChromiumImpl
+#include "src/third_party/blink/renderer/core/frame/location.cc"
+#undef Location
+
+namespace blink {
+
+Location::Location(DOMWindow* dom_window) : Location_ChromiumImpl(dom_window) {}
+
+DOMStringList* Location::ancestorOrigins() const {
+  auto* filtered_origins = MakeGarbageCollected<DOMStringList>();
+  const auto* innermost_origin =
+      DomWindow()->GetFrame()->GetSecurityContext()->GetSecurityOrigin();
+
+  auto* raw_origins = Location_ChromiumImpl::ancestorOrigins();
+  for (uint32_t i = 0; i < raw_origins->length(); i++) {
+    auto origin = SecurityOrigin::CreateFromString(raw_origins->item(i));
+    if (!origin->Host().EndsWith(".onion", kTextCaseASCIIInsensitive) ||
```

Comment:
i wonder if there are ways to bypass this, ex: punycode-encoding `.onion` or maybe using `.onion.` instead. not sure how much `origin->Host()` is normalized.

Reply:
I don't believe that IDNs are supported for Onion services since the name is not actually user-controlled. But also, the `.onion` suffix is fixed and ASCII, and the punycode for an ASCII name is the same as the ASCII name, at least according to my basic testing using https://www.punycoder.com/.

Reply:
`*.onion.` doesn't appear to resolve at all: http://ixrdj3iwwhkuau5tby5jh3a536a2rdhpbdbu6ldhng43r47kim7a3lid.onion./referrer/onion.html